Makes kube-proxy flags take precedence on proxy config

hanlins opened the pull request:
Signed-off-by: Hanlin Shi <shihanlin9@gmail.com>

<!--  Thanks for sending a pull request!  Here are some tips for you:

1. If this is your first time, please read our contributor guidelines: https://git.k8s.io/community/contributors/guide/first-contribution.md#your-first-contribution and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. Please label this pull request according to what type of issue you are addressing, especially if this is a release targeted pull request. For reference on required PR/issue labels, read here:
https://git.k8s.io/community/contributors/devel/sig-release/release.md#issuepr-kind-label
3. Ensure you have added or ran the appropriate tests for your PR: https://git.k8s.io/community/contributors/devel/sig-testing/testing.md
4. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/guide/pull-requests.md#best-practices-for-faster-reviews
5. If the PR is unfinished, see how to mark it: https://git.k8s.io/community/contributors/guide/pull-requests.md#marking-unfinished-pull-requests
-->

#### What type of PR is this?
/kind bug
<!--
Add one of the following kinds:
/kind bug
/kind cleanup
/kind documentation
/kind feature
/kind design

Optionally add one or more of the following kinds if applicable:
/kind api-change
/kind deprecation
/kind failing-test
/kind flake
/kind regression
-->

#### What this PR does / why we need it:
Kube-proxy flags are silently ignored when `--config` is set. This PR makes flags take precedence on the proxy component configuration.

#### Which issue(s) this PR fixes:
<!--
*Automatically closes linked issue when PR is merged.
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
_If PR is about `failing-tests or flakes`, please post the related issues/tests in a comment and do not use `Fixes`_*
-->
Fixes #98302 

#### Special notes for your reviewer:

#### Does this PR introduce a user-facing change?
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".

For more information on release notes see: https://git.k8s.io/community/contributors/guide/release-notes.md
-->
```release-note
Make kube-proxy flags take precedence on kube-proxy component configuration.
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

<!--
This section can be blank if this pull request does not require a release note.

When adding links which point to resources within git repositories, like
KEPs or supporting documentation, please reference a specific commit and avoid
linking directly to the master branch. This ensures that links reference a
specific point in time, rather than a document that may change over time.

See here for guidance on getting permanent links to files: https://help.github.com/en/articles/getting-permanent-links-to-files

Please use the following format for linking documentation:
- [KEP]: <link>
- [Usage]: <link>
- [Other doc]: <link>
-->
```docs

```

hanlins committed:
Makes kube-proxy flags take precedence on proxy config

Signed-off-by: Hanlin Shi <shihanlin9@gmail.com>

hanlins commented:
/cc @knabben

hanlins committed:
Update vendoring for mergo

Signed-off-by: Hanlin Shi <shihanlin9@gmail.com>

hanlins commented:
/area kube-proxy

hanlins commented:
/test pull-kubernetes-node-e2e

knabben commented:
/lgtm
/retest

knabben commented on cmd/kube-proxy/app/server_test.go:
```diff
@@ -625,3 +625,76 @@ func TestAddressFromDeprecatedFlags(t *testing.T) {
 	}
 }
+
+func TestFlagOverwrites(t *testing.T) {
```
Don't need to be on this PR but we can extend the testing for platform and instances specific ones.

hanlins replied:
Thanks @knabben, good point! I'll add coverage once reviewers reach a consensus on this overwriting behavior.

hanlins replied:
Thanks @knabben, good point! I'll add coverage once reviewers reach a consensus on this overwriting behavior. The test for now is just to demo flags will take precedence on proxy component configs.

jayunit100 commented:
@thockin this is the thing you were looking for, `mergo`, good job discovering it @hanlins !

liggitt commented on cmd/kube-proxy/app/server.go:
```diff
@@ -228,6 +229,8 @@ func (o *Options) Complete() error {
 			return err
 		}
+		// make flags take precedence
+		mergo.Merge(c, o.config, mergo.WithOverride)
```
does this make defaulted values override explicit config values?

hanlins replied:
No, have a test case here: https://github.com/kubernetes/kubernetes/blob/2e112f2efe15907af9b2dc2e485e74bc980df429/cmd/kube-proxy/app/server_test.go#L663-L668
You can see if the flag is not set but proxy component config is set, after the merge the value will be the one in proxy component config.

liggitt replied:
the unit test that makes sure this works doesn't match how Options are actually used by kube-proxy

kube-proxy:
1. creates options: `opts := NewOptions()`
2. applies defaults: `opts.config, err = opts.ApplyDefaults(opts.config)`
3. binds flags

I'm pretty sure the defaults in step 2 will stomp any config value specified with this current PR.

In the unit test, make sure you are exercising this the same way, and make sure the config file is effective when setting a value that is defaulted by opts.ApplyDefaults

liggitt replied:
> I'm pretty sure the defaults in step 2 will stomp any config value specified with this current PR.

If that turns out to be the case, that means our testing of the config file path has gaps that need to be closed as well

hanlins replied:
I see, that's a good point. I think the tricky part is it's hard to determine whether the value is from the default or explicitly passed from flags. The merge logic here is to make the values in `o.config` take precedence if it's a non-empty value, which can be our defaults.
I think one way to fix it is to move all default values for flags to a default proxy config struct, and merge it later without overwriting.

liggitt replied:
> I think one way to fix it is to move all default values for flags to a default proxy config struct, and merge it later without overwriting.

That makes the help displayed for the flags much less helpful, since it doesn't show the defaults used if no flags are passed. Before restructuring defaulting, we should determine whether this change is the right direction, see the comment in https://github.com/kubernetes/kubernetes/pull/101273#issuecomment-823357862

hanlins commented:
> @thockin this is the thing you were looking for, `mergo`, good job discovering it @hanlins !

Credits to @aojea!

liggitt commented:
This takes kube-proxy in a different direction than other components (kube-scheduler explicitly makes config and flags exclusive, with the exception of a few deprecated flags).

Whether or not flags should override config generally has been long-debated (cc @mtaufen @deads2k @sttts). I am of the opinion they should not, since it makes inspection/validation of a config file in isolation difficult to understand without also having the full invocation it is a part of.

hanlins commented:
Hi @liggitt, there're lots of back and forth on this flag precedence behavior, and I agree with your points. But one key difference between components like `kube-scheduler` and `kube-proxy` is `kube-proxy` sometimes needs instance-specific configurations, so in this sense, it's more similar to `kubelet` rather than `kube-scheduler`. Originally we plan to just use `--config` and `--instance-config` to cover both cases, but as you know https://github.com/kubernetes/enhancements/pull/1439 has been out for a while, but it still has some gaps to fill and could take a few more releases to get there. 
Before that instance-specific config happens, the only thing user have for per-instance config is the flags, but most of the flags are just silently ignored, it's like we want to provide people a better semantics for configurations in the future, we gave them one which is not ideal and the one we gave is not working. Last time during the WG meeting @thockin mentioned that we want to make our existing flag users happy, and we want to do flag overwriting similar to `kubelet` if the cost is not too large. So we take this direction on fixing the flag overwriting for our existing users.

liggitt commented:
> we want to do flag overwriting similar to `kubelet`

Having reviewed the kubelet flag/config merging changes originally, and many fixup changes over the years, I would strongly recommend against that path. The kubelet config initialization path is almost incomprehensible.

jayunit100 commented:
hmm... so, @liggitt are you suggesting we eliminate these flags entirely ?

liggitt commented:
"fail if config + flags overlap" avoids adding complex new merge behavior, and avoids the footgun of ignored flags. I'd start with that.

hanlins committed:
Add method to detect flag and config conflict in kube-proxy

1. Use struct tag to map config fields to flags.
2. Reject configration if conflict from config file and flags is
   detected.

Signed-off-by: Hanlin Shi <shihanlin9@gmail.com>

hanlins committed:
Add unit tests for kube-proxy config conflict detection

Signed-off-by: Hanlin Shi <shihanlin9@gmail.com>

hanlins commented:
/test pull-kubernetes-e2e-kind-ipv6

hanlins commented:
Hi @liggitt @knabben, made some updates here. I have to add some struct tag to associate the fields in `KubeProxyConfiguration` with the flags people use. With the change, we can detect and fast fail if the user sets the field in kube-proxy config and also sets the flag for it. But, to address the issue where people's flags got ignored silently, we still need to aggregate the configs from the config file and flags to one place, but since we've already checked that there's no conflict between config file and flags, we can safely combine the two without worry about the precedence. Any opinion on that?

liggitt commented:
> With the change, we can detect and fast fail if the user sets the field in kube-proxy config and also sets the flag for it.

What's the benefit to letting them set via config file field and flag but only when they match?

How does this determine the difference between a default value in the config and an explicitly set value in the config?

liggitt commented on staging/src/k8s.io/component-base/config/types.go:
```diff
@@ -23,17 +23,17 @@ import (
 // ClientConnectionConfiguration contains details for constructing a client.
 type ClientConnectionConfiguration struct {
```
This type is shared between multiple components (including out-of-tree components), so these flags are determined by the consuming component.

hanlins replied:
I see. I checked other components like kubelet, and they used the same flag for the same field, but I wasn't aware that maybe some external components are also consuming them. Would it be better if I mark make the struct tag something more specific to kube-proxy? Like change the tag from `flag` to `kubeProxyFlag`?

hanlins commented:
> What's the benefit to letting them set via config file field and flag but only when they match?
> 
> How does this determine the difference between a default value in the config and an explicitly set value in the config?

Actually, if the config is set in both the config file and the flag, no matter if they have the same value or not, we will consider this as a conflict and fail the command. In the test I have I'm using different values but actually it doesn't matter.

In our detection logic, we're actually detecting if the flag value is explicitly set by the user, or it's something from the flag default https://github.com/kubernetes/kubernetes/blob/9a9960d886dac65fc0cba17b99d50c3e546a5779/cmd/kube-proxy/app/server.go#L478
Basically, it's using this [Changed](https://pkg.go.dev/github.com/spf13/pflag#FlagSet.Changed) interface to detect whether the flag is set by user or not.

mtaufen commented:
> Whether or not flags should override config generally has been long-debated (cc @mtaufen @deads2k @sttts). I am of the opinion they should not, since it makes inspection/validation of a config file in isolation difficult to understand without also having the full invocation it is a part of.

I'm in agreement with this. We did it in kubelet because we were pursuing an incremental migration across releases and were trying to keep things backwards compatible when a flag later became available in config, but might still be set on the command line in an existing configuration. But if you can keep it either-or, the implementation is much simpler.

hanlins committed:
Amend conflict detection change

1. Add test to detect both flag and config file are set with same value.
2. Refine checking whether flag is using default value or being
   explicitly set by users.

Signed-off-by: Hanlin Shi <shihanlin9@gmail.com>

hanlins commented:
/test pull-kubernetes-e2e-kind-ipv6
